weld/subsystem/src/main/java/org/jboss/as/weld/WeldStartCompletionService.java

```diff
@@ -31,32 +38,73 @@
 import org.wildfly.security.manager.WildFlySecurityManager;
 
 /**
- * This service calls {@link Bootstrap#endInitialization()}, i.e. places the container into a state where it can service requests. It is started after all EE
- * components are installed which allows Weld to perform efficient cleanup and further optimizations after bootstrap.
+ * This service calls {@link Bootstrap#endInitialization()}, i.e. places the container into a state where it can service
+ * requests.
+ *
+ * Its start is delayed after all EE components are installed which allows Weld to perform efficient cleanup and further
+ * optimizations after bootstrap.
  *
  * @author Martin Kouba
+ * @author Matej Novotny
  * @see WeldStartService
  */
 public class WeldStartCompletionService implements Service<WeldStartCompletionService> {
 
     public static final ServiceName SERVICE_NAME = ServiceNames.WELD_START_COMPLETION_SERVICE_NAME;
 
     private final InjectedValue<WeldBootstrapService> bootstrap = new InjectedValue<WeldBootstrapService>();
+    private final InjectedValue<ExecutorService> executorService = new InjectedValue<ExecutorService>();
 
     private final ClassLoader classLoader;
+    private final List<ServiceController> serviceControllers;
 
-    public WeldStartCompletionService(ClassLoader classLoader) {
+    private final AtomicBoolean runOnce = new AtomicBoolean();
+
+    public WeldStartCompletionService(ClassLoader classLoader, List<ServiceController> serviceControllers) {
         this.classLoader = classLoader;
+        this.serviceControllers = serviceControllers;
     }
 
     @Override
     public void start(final StartContext context) throws StartException {
-        ClassLoader oldTccl = WildFlySecurityManager.getCurrentContextClassLoaderPrivileged();
+        if (!runOnce.compareAndSet(false, true)) {
+            // we only execute this once, if there is a restart, WeldStartService initiates re-deploy hence we do nothing here
+            return;
+        }
+
+        final ExecutorService executor = executorService.getValue();
+        final Runnable task = new Runnable() {
+            // we want to execute StabilityMonitor.awaitStability() in a different thread so that this is non-blocking
+            @Override
+            public void run() {
+                StabilityMonitor monitor = new StabilityMonitor();
+                for (ServiceController controller : serviceControllers) {
+                    monitor.addController(controller);
+                }
+                try {
+                    monitor.awaitStability();
+                } catch (InterruptedException ex) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(ex);
```

Comment:
You need to make sure that you always call either StartContext.complete() or StartContext.failed(), otherwise the deployment can hang.

Reply:
Ah, good catch, thanks!
Fixed.